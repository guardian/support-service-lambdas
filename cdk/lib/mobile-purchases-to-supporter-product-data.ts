import type { App } from 'aws-cdk-lib';
import { Duration } from 'aws-cdk-lib';
import { EventBus, Rule } from 'aws-cdk-lib/aws-events';
import { SqsQueue } from 'aws-cdk-lib/aws-events-targets';
import { AllowSupporterProductDataPutItemPolicy } from './cdk/policies';
import { SrSqsLambda } from './cdk/SrSqsLambda';
import { SrStack, type SrStageNames } from './cdk/SrStack';

export class MobilePurchasesToSupporterProductData extends SrStack {
	constructor(scope: App, stage: SrStageNames) {
		super(scope, { stage, app: 'mobile-purchases-to-supporter-product-data' });

		const errorImpact =
			'Supporter Product Data table is not being updated with mobile purchase data, which could result in users not receiving the correct benefits associated with their in app purchase.';
		const lambda = new SrSqsLambda(this, 'Lambda', {
			monitoring: { errorImpact },
			lambdaOverrides: {
				description:
					'A lambda that writes mobile purchase products into the supporter product data table',
				timeout: Duration.minutes(5),
			},
			maxReceiveCount: 3,
			// This must be >= the lambda timeout
			visibilityTimeout: Duration.minutes(5),
		});
		lambda.addPolicies(new AllowSupporterProductDataPutItemPolicy(this));

		const mobilePurchasesBus = EventBus.fromEventBusName(
			this,
			'MobilePurchasesBus',
			`mobile-purchases-dynamo-stream-${this.stage}`,
		);

		const forwardMobilePurchasesRule = new Rule(
			this,
			'ForwardMobilePurchasesToSupporterProductData',
			{
				ruleName: `forward-mobile-purchases-to-sqs-${this.stage}`,
				description:
					'Forward DynamoDB stream purchase events to mobile-purchases-to-supporter-product-data-lambda',
				eventBus: mobilePurchasesBus,
				eventPattern: {
					detail: {
						eventSource: ['aws:dynamodb'],
					},
				},
			},
		);

		forwardMobilePurchasesRule.addTarget(new SqsQueue(lambda.inputQueue));
	}
}
