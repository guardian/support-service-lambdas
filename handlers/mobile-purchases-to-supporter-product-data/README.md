# generate-product-catalog lambda

This lambda takes a DynamoDB stream of changes to the `user-subscriptions` table in the `gu-aws-mobile` account 
and updates the `SupporterProductData` account.
