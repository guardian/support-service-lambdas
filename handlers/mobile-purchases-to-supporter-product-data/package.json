{
  "name": "mobile-purchases-to-supporter-product-data",
  "scripts": {
    "cdk:test": "pnpm --filter cdk test mobile-purchases-to-supporter-product-data",
    "cdk:test-update": "pnpm --filter cdk test-update mobile-purchases-to-supporter-product-data",
    "test": "jest --group=-integration",
    "it-test": "jest --group=integration",
    "type-check": "tsc --noEmit",
    "build": "esbuild --bundle --platform=node --target=node20 --outdir=target/ src/index.ts --sourcemap",
    "lint": "eslint --cache --cache-location /tmp/eslintcache/ src/**/*.ts test/**/*.ts",
    "package": "pnpm type-check && pnpm lint && pnpm check-formatting && pnpm test && pnpm build && cd target && zip -qr mobile-purchases-to-supporter-product-data.zip ./*.js.map ./*.js",
    "check-formatting": "prettier --check \"**/*.ts\"",
    "fix-formatting": "prettier --write \"**/*.ts\"",
    "update-lambda": "../../update-lambda.sh \"mobile-purchases-to-supporter-product-data\"",
    "update-stack": "../../update-stack.sh \"mobile-purchases-to-supporter-product-data\"",
    "runFullSync": "ts-node  -r tsconfig-paths/register --project ../../tsconfig.json src/fullSyncCommand.ts"
  },
  "NOTICE1": "MANAGED FILE: to push changes see buildcheck/README.md - template: buildcheck/data/templates/handler/package.json.ts",
  "NOTICE2": "all dependencies are defined in buildcheck/data/build.ts",
  "dependencies": {
    "@aws-sdk/client-dynamodb": "^3.848.0",
    "zod": "catalog:",
    "dayjs": "^1.11.13"
  },
  "devDependencies": {
    "@types/aws-lambda": "^8.10.157",
    "csv-parse": "^5.6.0",
    "ts-node": "^10.9.1",
    "tsconfig-paths": "catalog:"
  }
}
