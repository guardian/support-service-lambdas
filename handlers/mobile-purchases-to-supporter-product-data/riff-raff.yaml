# MANAGED FILE: to push changes see buildcheck/README.md - template: buildcheck/data/templates/handler/riff-raff.yaml.ts
stacks:
  - support
regions:
  - eu-west-1
allowedStages:
  - CODE
  - PROD
deployments:
  mobile-purchases-to-supporter-product-data-cloudformation:
    type: cloud-formation
    app: mobile-purchases-to-supporter-product-data
    parameters:
      templateStagePaths:
        CODE: mobile-purchases-to-supporter-product-data-CODE.template.json
        PROD: mobile-purchases-to-supporter-product-data-PROD.template.json
  mobile-purchases-to-supporter-product-data:
    type: aws-lambda
    parameters:
      fileName: mobile-purchases-to-supporter-product-data.zip
      bucketSsmLookup: true
      prefixStack: false
      functionNames:
        - mobile-purchases-to-supporter-product-data-
    dependencies:
      - mobile-purchases-to-supporter-product-data-cloudformation
