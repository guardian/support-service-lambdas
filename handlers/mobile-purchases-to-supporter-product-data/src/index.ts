import { logger } from '@modules/routing/logger';
import type { Stage } from '@modules/stage';
import { stageFromEnvironment } from '@modules/stage';
import type { DynamoDBStreamEvent, Handler } from 'aws-lambda';

const updateSupporterProductData = (
	stage: Stage,
	event: DynamoDBStreamEvent,
) => {
	logger.log(
		'info',
		`Updated supporter product data for stage ${stage} with event ${JSON.stringify(event)}`,
	);
	return Promise.resolve();
};
export const handler: Handler = async (event: DynamoDBStreamEvent) => {
	console.log(`Input is ${JSON.stringify(event, null, 2)}`);
	const stage = stageFromEnvironment();
	await updateSupporterProductData(stage, event);
};
