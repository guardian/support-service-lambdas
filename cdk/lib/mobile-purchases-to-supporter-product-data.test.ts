import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { MobilePurchasesToSupporterProductData } from './mobile-purchases-to-supporter-product-data';

describe('The MobilePurchasesToSupporterProductData stack', () => {
	it('matches the snapshot', () => {
		const app = new App();
		const codeStack = new MobilePurchasesToSupporterProductData(app, 'CODE');
		const prodStack = new MobilePurchasesToSupporterProductData(app, 'PROD');

		expect(Template.fromStack(codeStack).toJSON()).toMatchSnapshot();
		expect(Template.fromStack(prodStack).toJSON()).toMatchSnapshot();
	});
});
