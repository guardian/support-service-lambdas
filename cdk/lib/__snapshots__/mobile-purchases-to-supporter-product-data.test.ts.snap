// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The MobilePurchasesToSupporterProductData stack matches the snapshot 1`] = `
{
  "Metadata": {
    "gu:cdk:constructs": [
      "GuDistributionBucketParameter",
    ],
    "gu:cdk:version": "TEST",
  },
  "Parameters": {
    "DistributionBucketName": {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "ForwardMobilePurchasesToSupporterProductDataAllowEventRulemobilepurchasestosupporterproductdataCODEmobilepurchasestosupporterproductdatalambda8483822015D451A7": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "mobilepurchasestosupporterproductdatalambdaC3CC84DF",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": {
          "Fn::GetAtt": [
            "ForwardMobilePurchasesToSupporterProductDataBA6447A8",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "ForwardMobilePurchasesToSupporterProductDataBA6447A8": {
      "Properties": {
        "Description": "Forward DynamoDB stream purchase events to supporter product data lambda",
        "EventBusName": "mobile-purchases-dynamo-stream-CODE",
        "EventPattern": {
          "detail-type": [
            "DynamoDB Streams Record",
          ],
          "source": [
            "aws.dynamodb",
          ],
        },
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Fn::GetAtt": [
                "mobilepurchasestosupporterproductdatalambdaC3CC84DF",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "mobilepurchasestosupporterproductdatalambdaC3CC84DF": {
      "DependsOn": [
        "mobilepurchasestosupporterproductdatalambdaServiceRoleDefaultPolicy660F159E",
        "mobilepurchasestosupporterproductdatalambdaServiceRole499F7137",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "DistributionBucketName",
          },
          "S3Key": "support/CODE/mobile-purchases-to-supporter-product-data/mobile-purchases-to-supporter-product-data.zip",
        },
        "Description": "A lambda that writes mobile purchase products into the supporter product data table",
        "Environment": {
          "Variables": {
            "APP": "mobile-purchases-to-supporter-product-data",
            "NODE_OPTIONS": "--enable-source-maps",
            "STACK": "support",
            "STAGE": "CODE",
          },
        },
        "FunctionName": "mobile-purchases-to-supporter-product-data-CODE",
        "Handler": "index.handler",
        "LoggingConfig": {
          "LogFormat": "Text",
        },
        "MemorySize": 1024,
        "Role": {
          "Fn::GetAtt": [
            "mobilepurchasestosupporterproductdatalambdaServiceRole499F7137",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Tags": [
          {
            "Key": "App",
            "Value": "mobile-purchases-to-supporter-product-data",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-service-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "support",
          },
          {
            "Key": "Stage",
            "Value": "CODE",
          },
        ],
        "Timeout": 15,
      },
      "Type": "AWS::Lambda::Function",
    },
    "mobilepurchasestosupporterproductdatalambdaServiceRole499F7137": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": [
          {
            "Key": "App",
            "Value": "mobile-purchases-to-supporter-product-data",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-service-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "support",
          },
          {
            "Key": "Stage",
            "Value": "CODE",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "mobilepurchasestosupporterproductdatalambdaServiceRoleDefaultPolicy660F159E": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                      "/support/CODE/mobile-purchases-to-supporter-product-data/mobile-purchases-to-supporter-product-data.zip",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/CODE/support/mobile-purchases-to-supporter-product-data",
                  ],
                ],
              },
            },
            {
              "Action": [
                "ssm:GetParameters",
                "ssm:GetParameter",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/CODE/support/mobile-purchases-to-supporter-product-data/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "mobilepurchasestosupporterproductdatalambdaServiceRoleDefaultPolicy660F159E",
        "Roles": [
          {
            "Ref": "mobilepurchasestosupporterproductdatalambdaServiceRole499F7137",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;

exports[`The MobilePurchasesToSupporterProductData stack matches the snapshot 2`] = `
{
  "Metadata": {
    "gu:cdk:constructs": [
      "GuDistributionBucketParameter",
    ],
    "gu:cdk:version": "TEST",
  },
  "Parameters": {
    "DistributionBucketName": {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "ForwardMobilePurchasesToSupporterProductDataAllowEventRulemobilepurchasestosupporterproductdataPRODmobilepurchasestosupporterproductdatalambdaF9674256F2C10C35": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "mobilepurchasestosupporterproductdatalambdaC3CC84DF",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": {
          "Fn::GetAtt": [
            "ForwardMobilePurchasesToSupporterProductDataBA6447A8",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "ForwardMobilePurchasesToSupporterProductDataBA6447A8": {
      "Properties": {
        "Description": "Forward DynamoDB stream purchase events to supporter product data lambda",
        "EventBusName": "mobile-purchases-dynamo-stream-PROD",
        "EventPattern": {
          "detail-type": [
            "DynamoDB Streams Record",
          ],
          "source": [
            "aws.dynamodb",
          ],
        },
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Fn::GetAtt": [
                "mobilepurchasestosupporterproductdatalambdaC3CC84DF",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "mobilepurchasestosupporterproductdatalambdaC3CC84DF": {
      "DependsOn": [
        "mobilepurchasestosupporterproductdatalambdaServiceRoleDefaultPolicy660F159E",
        "mobilepurchasestosupporterproductdatalambdaServiceRole499F7137",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "DistributionBucketName",
          },
          "S3Key": "support/PROD/mobile-purchases-to-supporter-product-data/mobile-purchases-to-supporter-product-data.zip",
        },
        "Description": "A lambda that writes mobile purchase products into the supporter product data table",
        "Environment": {
          "Variables": {
            "APP": "mobile-purchases-to-supporter-product-data",
            "NODE_OPTIONS": "--enable-source-maps",
            "STACK": "support",
            "STAGE": "PROD",
          },
        },
        "FunctionName": "mobile-purchases-to-supporter-product-data-PROD",
        "Handler": "index.handler",
        "LoggingConfig": {
          "LogFormat": "Text",
        },
        "MemorySize": 1024,
        "Role": {
          "Fn::GetAtt": [
            "mobilepurchasestosupporterproductdatalambdaServiceRole499F7137",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Tags": [
          {
            "Key": "App",
            "Value": "mobile-purchases-to-supporter-product-data",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-service-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "support",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
        "Timeout": 15,
      },
      "Type": "AWS::Lambda::Function",
    },
    "mobilepurchasestosupporterproductdatalambdaServiceRole499F7137": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": [
          {
            "Key": "App",
            "Value": "mobile-purchases-to-supporter-product-data",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-service-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "support",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "mobilepurchasestosupporterproductdatalambdaServiceRoleDefaultPolicy660F159E": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                      "/support/PROD/mobile-purchases-to-supporter-product-data/mobile-purchases-to-supporter-product-data.zip",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/PROD/support/mobile-purchases-to-supporter-product-data",
                  ],
                ],
              },
            },
            {
              "Action": [
                "ssm:GetParameters",
                "ssm:GetParameter",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/PROD/support/mobile-purchases-to-supporter-product-data/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "mobilepurchasestosupporterproductdatalambdaServiceRoleDefaultPolicy660F159E",
        "Roles": [
          {
            "Ref": "mobilepurchasestosupporterproductdatalambdaServiceRole499F7137",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
